src/diagnostic/emitter.rs

```diff
@@ -0,0 +1,313 @@
+
+        let max_spaces = self.max_line_num_width(diag);
+        let mut is_first = true;
+
+        if let Some(primary) = diag.primary {
+            self.emit_line(&mut buffer, primary, None, true, max_spaces, diag.level);
+            is_first = false;
+        }
+
+        for (span, label) in diag.spans.iter() {
+            self.emit_line(
+                &mut buffer,
+                *span,
+                Some(label.clone()),
+                is_first,
+                max_spaces,
+                diag.level,
+            );
+            is_first = false;
+        }
+
+        for subd in diag.children.iter() {
+            if let Some(span) = subd.span {
+                buffer.puts(subd.level.as_styled_string());
+
+                buffer.puts(StyledString::new(
+                    format!(": {}\n", subd.message),
+                    super::styled::Style::MainHeaderMsg,
+                ));
+
+                self.emit_line(&mut buffer, span, None, true, max_spaces, subd.level);
+            } else {
+                buffer.puts(StyledString::new(
+                    format!("{:spaces$} = ", "", spaces = max_spaces),
+                    super::styled::Style::LineAndColumn,
+                ));
+
+                buffer.puts(StyledString::new(
+                    format!("{}: ", subd.level.to_str()),
+                    super::styled::Style::MainHeaderMsg,
+                ));
+
+                buffer.puts(StyledString::new(
+                    format!("{}\n", subd.message),
+                    super::styled::Style::NoStyle,
+                ));
+            }
+        }
+
+        buffer.puts(StyledString::new(
+            String::from("\n"),
+            super::styled::Style::NoStyle,
+        ));
+
+        // Render the buffer we have accumulated
+        self.render_buffer(&mut stream, &buffer)?;
+
+        Ok(())
+    }
+
+    fn render_buffer(
+        &self,
+        stream: &mut StandardStream,
+        buffer: &StyledBuffer,
+    ) -> Result<(), Error> {
+        for string in buffer.iter() {
+            let spec = string.style.to_spec();
+
+            if self.colored {
+                stream.set_color(&spec)?;
+            }
+
+            write!(stream, "{}", string.text)?;
+
+            if self.colored {
+                stream.set_color(&ColorSpec::new())?;
+            }
+        }
+
+        Ok(())
+    }
+
+    fn max_line_num_width(&self, diag: &Diagnostic) -> usize {
+        let mut spans = Vec::new();
+        let mut max_width = 0;
+
+        if let Some(primary) = diag.primary {
+            spans.push(primary);
+        }
+
+        // That could kind of be a lot of copies, is this a concern?
+        for (span, _) in diag.spans.iter() {
+            spans.push(*span);
+        }
+
+        for span in spans {
+            let source_file = if let Some(source_file) = self.source_manager.get_file(span.source) {
+                source_file
+            } else {
+                panic!(
```

Comment:
Also if we have a wrapper type that is not publicly construct-able, we can be pretty sure `get_file` will never fail as we control how that type is given out

Reply:
I added get_file_unwrap() for code readability purposes, although the panic should only ever be in those two places and nowhere else in the code unless a new part were added to the emitter. SourceManager.get_file() should only really be used by the emitter and other low-level machinery, but it is a good design decision as code goes.
Fixed in b9244695e0e806aba5aadaac8a4b720811edd3b1

Reply:
Sounds good!

Reply:
It is something that never should fail, but in theory it can and it has happened before while it was being written, so we shouldn't have an unsafe function or anything, but it need not be a long process to simply index into the source manager